modules/oe_content_event/src/Plugin/InternalLinkSourceFilter/EventPeriodFilter.php

```diff
@@ -0,0 +1,148 @@
+use Drupal\oe_time_caching\Cache\TimeBasedCacheTagGeneratorInterface;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Event period link source filter class.
+ *
+ * @InternalLinkSourceFilter(
+ *   id = "oe_content_event_period",
+ *   label = @Translation("Event period link source filter"),
+ *   description = @Translation("Filters events by the period they are in: past or upcoming."),
+ *   entity_types = {
+ *     "node" = {
+ *       "oe_event",
+ *     },
+ *   },
+ * )
+ */
+class EventPeriodFilter extends InternalLinkSourceFilterPluginBase implements InternalLinkSourceFilterInterface, ContainerFactoryPluginInterface {
+
+  /**
+   * Option for upcoming events.
+   */
+  const UPCOMING = 1;
+
+  /**
+   * Option for past events.
+   */
+  const PAST = -1;
+
+  /**
+   * The time based cache generator.
+   *
+   * @var \Drupal\oe_time_caching\Cache\TimeBasedCacheTagGeneratorInterface
+   */
+  protected $timeBasedCacheTagGenerator;
+
+  /**
+   * The system time.
+   *
+   * @var \Drupal\Component\Datetime\TimeInterface
+   */
+  protected $time;
+
+  /**
+   * {@inheritdoc}
+   */
+  public function defaultConfiguration() {
+    return [
+      'period' => self::UPCOMING,
+    ];
+  }
+
+  /**
+   * Constructs the EventPeriodFilter plugin.
+   *
+   * @param array $configuration
+   *   The plugin configuration.
+   * @param string $plugin_id
+   *   The plugin id.
+   * @param mixed $plugin_definition
+   *   The plugin definition.
+   * @param \Drupal\oe_time_caching\Cache\TimeBasedCacheTagGeneratorInterface $time_based_cache_tag_generator
+   *   The time based cache generator.
+   * @param \Drupal\Component\Datetime\TimeInterface $time
+   *   The system time.
+   */
+  public function __construct(array $configuration, $plugin_id, $plugin_definition, TimeBasedCacheTagGeneratorInterface $time_based_cache_tag_generator, TimeInterface $time) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+    $this->timeBasedCacheTagGenerator = $time_based_cache_tag_generator;
+    $this->time = $time;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+    return new static(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+      $container->get('oe_time_caching.time_based_cache_tag_generator'),
+      $container->get('datetime.time')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function apply(QueryInterface $query, array $context, RefinableCacheableDependencyInterface $cacheability): void {
+    $now = new DrupalDateTime();
+    $current_time = $this->time->getCurrentTime();
+    $now->setTimestamp($current_time);
+    $now->setTimezone(new \DateTimeZone(DateTimeItemInterface::STORAGE_TIMEZONE));
+    $cacheability->addCacheTags($this->timeBasedCacheTagGenerator->generateTags($now->getPhpDateTime()));
```

Comment:
Shouldn't we create caching tag based on the newest/oldest event node? Like that we are creating caching tags that are based on the first time a user visits a page where this is displayed.

Reply:
I'm not sure we need to test that caches are invalidated on cron, shouldn't that need to happen on oe_time_caching?

Reply:
Yes I meant that the oe_time_caching module will invalidate them on cron. But ok we can also not test it here as that's tested in the module already.